libs/shared/form-fields/src/lib/FileUploadController/index.tsx

```diff
@@ -0,0 +1,251 @@
+  readonly id: string
+  error?: string
+  application: Application
+  readonly header?: string
+  readonly description?: string
+  readonly buttonLabel?: string
+  readonly multiple?: boolean
+  readonly accept?: string
+  readonly maxSize?: number
+}
+
+export const FileUploadController: FC<FileUploadControllerProps> = ({
+  id,
+  error,
+  application,
+  header,
+  description,
+  buttonLabel,
+  multiple,
+  accept,
+  maxSize,
+}) => {
+  const { clearErrors, setValue } = useFormContext()
+  const [uploadError, setUploadError] = useState<string | undefined>(undefined)
+  const val = getValueViaPath(application.answers, id, []) as UploadFile[]
+
+  const [createUploadUrl] = useMutation(CREATE_UPLOAD_URL)
+
+  const [addAttachment] = useMutation(ADD_ATTACHMENT)
+
+  const [deleteAttachment] = useMutation(DELETE_ATTACHMENT)
+
+  const initialUploadFiles: UploadFile[] =
+    (val && val.map((f) => answerToUploadFile(f))) || []
+
+  const [state, dispatch] = useReducer(reducer, initialUploadFiles)
+
+  useEffect(() => {
+    const onlyUploadedFiles = state.filter(
+      (f: UploadFile) => f.key && f.url && f.status === 'done',
+    )
+
+    const uploadAnswer: UploadFileAnswer[] = onlyUploadedFiles.map(
+      transformToAnswer,
+    )
+
+    setValue(id, uploadAnswer)
+  }, [state, id, setValue])
+
+  const uploadFileFlow = async (file: UploadFile) => {
+    try {
+      // 1. Get the upload URL
+      const { data } = await createUploadUrl({
+        variables: {
+          filename: file.name,
+        },
+      })
+
+      // 2. Upload the file to S3
+      const {
+        createUploadUrl: { url, fields },
+      } = data
+
+      const response = await uploadFileToS3(file, dispatch, url, fields)
+
+      // 3. Add Attachment Data
+      await addAttachment({
+        variables: {
+          input: {
+            id: application.id,
+            key: fields.key,
+            url: `${response.url}/${fields.key}`,
+          },
+        },
+      })
+
+      // Done!
+      return Promise.resolve({ url: response.url, key: fields.key })
+    } catch (e) {
+      error = e
+      return Promise.reject(e)
+    }
+  }
+
+  const onFileChange = async (newFiles: File[]) => {
+    const addedUniqueFiles = newFiles.filter((newFile: File) => {
+      let isUnique = true
+      state.forEach((uploadedFile: UploadFile) => {
+        if (uploadedFile.name === newFile.name) isUnique = false
+      })
+      return isUnique
+    })
+
+    if (addedUniqueFiles.length === 0) return
+
+    clearErrors(id)
+    setUploadError(undefined)
+
+    const newUploadFiles = addedUniqueFiles.map((f) =>
+      fileToObject(f, 'uploading'),
+    )
+
+    // Add the files to the list so that the control presents them
+    // with a spinner.
+    dispatch({
+      type: ActionTypes.ADD,
+      payload: {
+        newFiles: newUploadFiles,
+      },
+    })
+
+    // Upload each file.
+    newUploadFiles.forEach(async (f: UploadFile) => {
```

Comment:
Using async/await and then/catch together is a bit weird. I would probably do

```typescript
Promise.all(newUploadFiles.map(async (f: UploadFile) => {
  try {
    const res = await uploadFileFlow(f);
    
    dispatch({
      type: ActionTypes.UPDATE,
      payload: {
        file: f,
        status: 'done',
        percent: 100,
        key: res.key,
        url: res.url,
      },
    })
  } catch (e) {
    ...
  }
}))
```

Reply:
Good point thanks, updated now.